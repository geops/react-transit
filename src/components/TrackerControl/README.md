#

This demonstrates the use of RouteSchedule.

```jsx
import React from 'react';
import Dialog from 'react-spatial/components/Dialog';
import BasicMap from 'react-spatial/components/BasicMap';
import Layer from 'react-spatial/Layer';
import TileLayer from 'ol/layer/Tile';
import OSMSource from 'ol/source/OSM';
import TrajservLayer from 'react-transit/layers/TrajservLayer';
import RouteSchedule from 'react-transit/components/RouteSchedule';

const trackerLayer = new TrajservLayer();
const layers = [
  new Layer({
    name: 'Layer',
    olLayer: new TileLayer({
      source: new OSMSource(),
    }),
  }),
  trackerLayer
]

function  TrackerControlExample()  {
  return (
    <>
      <BasicMap center={[951560, 6002550]} zoom={14} layers={layers}/>
      <TrackerControl onChange={speed => trackerLayer.setSpeed(speed)}>
      </TrackerControl>
    </>
  );
}

<TrackerControlExample/>;
```
