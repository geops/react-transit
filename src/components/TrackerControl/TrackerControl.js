import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { FaPlay, FaForward, FaBackward } from 'react-icons/fa';

const increaseSpeed = speed => {
  let delta = 0.1;
  if (speed >= 1) {
    delta = 1;
  }
  if (speed >= 10) {
    delta = 5;
  }
  const nextSpeed = speed + delta;
  return nextSpeed > 30 ? speed : nextSpeed;
};

const decreaseSpeed = speed => {
  let delta = 0.1;
  if (speed > 1) {
    delta = 1;
  }
  if (speed > 10) {
    delta = 5;
  }
  const nextSpeed = speed - delta;
  if (nextSpeed < 0.1) {
    return speed;
  }
  return nextSpeed;
};

const defaultRenderButton = (icon, onClick) => {
  return (
    <button type="button" onClick={onClick}>
      {icon}
    </button>
  );
};

function TrackerControl({
  iconSpeedDown,
  iconSpeedReset,
  iconSpeedUp,
  renderButton,
  onChange,
}) {
  const [speed, setSpeed] = useState(1);

  useEffect(() => {
    onChange(speed);
  }, [speed, onChange]);

  return (
    <>
      {renderButton(iconSpeedDown, () => setSpeed(decreaseSpeed(speed)))}
      {renderButton(iconSpeedReset, () => setSpeed(1))}
      {renderButton(iconSpeedUp, () => setSpeed(increaseSpeed(speed)))}
      <span>{`${speed < 1 ? speed.toFixed(1) : speed}x`}</span>
    </>
  );
}

TrackerControl.propTypes = {
  iconSpeedDown: PropTypes.element,
  iconSpeedUp: PropTypes.element,
  iconSpeedReset: PropTypes.element,
  renderButton: PropTypes.func,
  onChange: PropTypes.func,
};

TrackerControl.defaultProps = {
  iconSpeedDown: <FaBackward />,
  iconSpeedReset: <FaPlay />,
  iconSpeedUp: <FaForward />,
  renderButton: defaultRenderButton,
  onChange: () => {},
};

export default TrackerControl;
