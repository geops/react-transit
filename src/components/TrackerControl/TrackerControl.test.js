import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TrackerControl from '.';

configure({ adapter: new Adapter() });

describe('TrackerControl', () => {
  test('matches snapshots.', () => {
    const component = renderer.create(<TrackerControl />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
