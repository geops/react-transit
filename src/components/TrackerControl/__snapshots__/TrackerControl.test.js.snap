// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackerControl matches snapshots. 1`] = `
Array [
  <button
    onClick={[Function]}
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.5 280.6l192 160c20.6 17.2 52.5 2.8 52.5-24.6V96c0-27.4-31.9-41.8-52.5-24.6l-192 160c-15.3 12.8-15.3 36.4 0 49.2zm256 0l192 160c20.6 17.2 52.5 2.8 52.5-24.6V96c0-27.4-31.9-41.8-52.5-24.6l-192 160c-15.3 12.8-15.3 36.4 0 49.2z"
      />
    </svg>
  </button>,
  <button
    onClick={[Function]}
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
      />
    </svg>
  </button>,
  <button
    onClick={[Function]}
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M500.5 231.4l-192-160C287.9 54.3 256 68.6 256 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2zm-256 0l-192-160C31.9 54.3 0 68.6 0 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2z"
      />
    </svg>
  </button>,
  <span>
    1x
  </span>,
]
`;
